File: DataStructures/Trees/BinaryIndexedTree/Other/Add one BIT point 2/Add-one-BIT-point-GPT-2.py
# 以下に、**Python 3.8.10** 向けに、**BIT (Binary Indexed Tree / フェニック木)** を使って **一点更新・クエリ後のBIT状態出力** を行うコードを示します。

# ---

# ## ✅ 実装要件

# * **型ヒント付き**
# * 処理を関数化
# * 入力は `sys.stdin`（`fs` 相当）
# * メモリと処理時間を考慮（最大で n, Q ≤ 500）

# ---

# ## ✅ Python実装（コメント付き）

# ```python
import sys
from typing import List


class BIT:
    """Binary Indexed Tree (Fenwick Tree) for cumulative sum updates."""

    def __init__(self, size: int) -> None:
        self.n: int = size
        self.tree: List[int] = [0] * (self.n + 1)  # 1-indexed

    def add(self, index: int, value: int) -> None:
        """
        Adds `value` to the element at position `index`.
        index: 1-based index
        Time Complexity: O(log n)
        """
        while index <= self.n:
            self.tree[index] += value
            index += index & -index

    def get_tree(self) -> List[int]:
        """
        Returns the current BIT tree with 0 at index 0 (for output).
        Time Complexity: O(n)
        """
        return [0] + self.tree[1:]  # Output format includes 0 at the beginning


def process_queries(n: int, A: List[int], queries: List[List[int]]) -> None:
    """
    Processes the update queries and prints the BIT after each.
    
    Parameters:
    - n: int - Size of the array A
    - A: List[int] - Initial array A[0..n-1]
    - queries: List[List[int]] - Each query is [I, b] representing A[I-1] += b
    
    Output:
    - For each query, print the current BIT state
    """

    bit: BIT = BIT(n)

    # Initial construction of BIT from A
    for i in range(n):
        bit.add(i + 1, A[i])  # BIT is 1-indexed

    for I, b in queries:
        bit.add(I, b)  # Update A[I-1] += b, reflect in BIT
        print(' '.join(map(str, bit.get_tree())))


def main() -> None:
    """
    Main function to read input and call the query processor.
    Assumes input from stdin.
    """
    input_lines = sys.stdin.read().strip().split('\n')
    n: int = int(input_lines[0])
    A: List[int] = list(map(int, input_lines[1].split()))
    Q: int = int(input_lines[2])
    queries: List[List[int]] = [list(map(int, line.split())) for line in input_lines[3:3 + Q]]

    process_queries(n, A, queries)


if __name__ == '__main__':
    main()
# ```

# ---

# ## ✅ 処理時間とメモリの見積もり

# | 項目      | 詳細                                  | 計算量               |
# | ------- | ----------------------------------- | ----------------- |
# | 初期BIT構築 | `n` 回の `add`                        | `O(n log n)`      |
# | クエリ処理   | `Q` 回の `add` と `get_tree`           | `O(Q log n + Qn)` |
# | メモリ使用量  | `A`: 4n bytes, `tree`: 4(n+1) bytes | 約 8KB (n=500)     |

# → 実行時間・メモリ消費は**非常に軽量**です（数百個まで）。

# ---

# ## ✅ 実行方法例

# ファイル `main.py` として保存後：

# ```bash
# python3 main.py < input.txt
# ```

# `input.txt` の内容：

# ```
# 6
# 1 5 7 9 8 6
# 5
# 5 4
# 1 10
# 3 9
# 3 8
# 4 7
# ```

# 出力：

# ```
# 0 1 6 7 22 12 18
# 0 11 16 7 32 12 18
# 0 11 16 16 41 12 18
# 0 11 16 24 49 12 18
# 0 11 16 24 56 12 18
# ```

# ---

# 必要であれば、`sum(i)` や `range_sum(l, r)` の機能や図解も追加できます。希望があればお知らせください。

File: DataStructures/Trees/BinaryIndexedTree/Other/Add one BIT point 2/Add-one-BIT-point-GPT-2.js
// 以下は、指定された条件（Node.js v16.17.1、`fs` を用いた標準入力処理、処理時間・メモリ消費に注意）に基づき、BIT（Binary Indexed Tree: フェニック木）を用いた一点更新処理の実装です。

// ---

// ### 🔧 実装方針

// * BIT（Fenwick Tree）を配列 A から構築。
// * 各クエリで `A[I] += b` を行い、**BIT構造の該当箇所**を更新。
// * 毎回 `BIT` を出力（インデックス0から）。

// ---

// ### ✅ コード（Node.js）

// ```javascript
const fs = require('fs');

/**
 * Binary Indexed Tree (Fenwick Tree) 構造体
 * @param {number} size - 配列の長さ
 */
class BIT {
    constructor(size) {
        this.n = size;
        this.tree = Array(this.n + 1).fill(0); // 1-indexed
    }

    /**
     * 値を加算（1-indexed）
     * @param {number} index - 加算対象のインデックス（1-based）
     * @param {number} value - 加算する値
     */
    add(index, value) {
        while (index <= this.n) {
            this.tree[index] += value;
            index += index & -index;
        }
    }

    /**
     * A_1 ~ A_index までの累積和を取得（1-indexed）
     * @param {number} index 
     * @returns {number} 累積和
     */
    sum(index) {
        let result = 0;
        while (index > 0) {
            result += this.tree[index];
            index -= index & -index;
        }
        return result;
    }

    /**
     * BIT全体の内部状態を取得
     * 0番目に0を入れた状態で返す（出力用）
     * @returns {number[]} BIT配列
     */
    getTree() {
        return [0, ...this.tree.slice(1)];
    }
}

// 標準入力読み込みと処理
function main(input) {
    const lines = input.trim().split('\n');
    const n = parseInt(lines[0]);
    const A = lines[1].split(' ').map(Number);
    const Q = parseInt(lines[2]);
    const queries = lines.slice(3).map(line => line.split(' ').map(Number));

    const bit = new BIT(n);

    // 初期BIT構築（1-indexed）
    for (let i = 0; i < n; i++) {
        bit.add(i + 1, A[i]);
    }

    for (let [I, b] of queries) {
        // A[I-1] に b を加算、BITに反映
        bit.add(I, b);

        // 現在のBITの状態出力（0からn番目まで）
        const tree = bit.getTree();
        console.log(tree.join(' '));
    }
}

// fs を用いた入力処理
const input = fs.readFileSync('/dev/stdin', 'utf8');
main(input);
// ```

// ---

// ### 💡 補足情報

// * `BIT.add(index, value)` は O(log n)
// * クエリは Q 個なので、全体計算量は O((n + Q) log n)
// * メモリ使用量：

//   * BIT: `O(n)`
//   * A, tree配列合わせても `~数KB` 程度（n, Q ≤ 500）

// ---

// ### ✅ 出力例確認（手動）

// 入力例:

// ```
// 6
// 1 5 7 9 8 6
// 5
// 5 4
// 1 10
// 3 9
// 3 8
// 4 7
// ```

// 出力（期待通り）:

// ```
// 0 1 6 7 22 12 18
// 0 11 16 7 32 12 18
// 0 11 16 16 41 12 18
// 0 11 16 24 49 12 18
// 0 11 16 24 56 12 18
// ```

// ---

// 必要があれば、経路図やBIT更新の詳細図解も追記可能です。
